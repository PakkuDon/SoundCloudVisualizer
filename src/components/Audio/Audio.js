import React, { useEffect, useRef } from "react"
import PropTypes from "prop-types"

function Audio(props) {
  const audioElement = useRef(null)

  useEffect(() => {
    const audioContext = new (window.AudioContext ||
      window.webkitAudioContext)()
    if (audioContext) {
      const source = audioContext.createMediaElementSource(audioElement.current)
      const analyser = audioContext.createAnalyser()
      source.connect(analyser)
      analyser.connect(audioContext.destination)
      analyser.fftSize = 256
      props.onRender(analyser)
    }
  }, [])

  const { onEnded, src } = props

  return (
    <audio
      ref={audioElement}
      src={src}
      autoPlay
      crossOrigin="anonymous"
      controls
      onEnded={onEnded}
    >
      Your browser does not support the HTML5 audio element.
    </audio>
  )
}

Audio.propTypes = {
  src: PropTypes.string,
  onEnded: PropTypes.func,
  onRender: PropTypes.func,
}

Audio.defaultProps = {
  onEnded: () => {},
  onRender: () => {},
}

export default Audio
