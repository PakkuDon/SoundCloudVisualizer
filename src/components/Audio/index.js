import React from 'react'
import PropTypes from 'prop-types'

class Audio extends React.Component {
  constructor(props) {
    super(props)
    this.audioRef = React.createRef()
  }

  componentDidMount() {
    const audioContext = new (window.AudioContext || window.webkitAudioContext)
    if (audioContext) {
      const source = audioContext.createMediaElementSource(this.audioRef.current)
      const analyser = audioContext.createAnalyser()
      source.connect(analyser)
      analyser.connect(audioContext.destination)
      this.props.onRender(analyser)
    }
  }

  render() {
    const { src } = this.props

    return (
      <audio ref={this.audioRef} src={src} autoPlay crossOrigin="anonymous" controls>
        Your browser does not support the HTML5 audio element.
      </audio>
    )
  }
}

Audio.propTypes = {
  src: PropTypes.string,
  onRender: PropTypes.func,
}

Audio.defaultProps = {
  onRender: () => {},
}

export default Audio
