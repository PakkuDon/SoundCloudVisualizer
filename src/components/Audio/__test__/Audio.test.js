import React from "react"
import { render, fireEvent } from "@testing-library/react"

import Audio from "../Audio"

describe("Audio", () => {
  beforeEach(() => {
    window.AudioContext = jest.fn().mockImplementation(() => {
      return {
        createMediaElementSource: () => ({ connect() {} }),
        createAnalyser: () => ({ connect() {} }),
      }
    })
  })

  afterEach(() => {
    window.AudioContext.mockRestore()
  })

  it("uses provided src", () => {
    const src = "https://example.com/example.mp3"
    const { container } = render(<Audio src={src} />)

    expect(container).toMatchSnapshot()
  })

  describe("on audio end", () => {
    it("calls supplied onEnded handler", () => {
      const spy = jest.fn()
      const { container } = render(<Audio onEnded={spy} />)

      fireEvent.ended(container.querySelector("audio"))

      expect(spy).toBeCalled()
    })
  })
})
