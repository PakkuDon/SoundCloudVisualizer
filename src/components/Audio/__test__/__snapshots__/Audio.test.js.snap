// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Audio uses provided src 1`] = `
<div>
  <audio
    autoplay=""
    controls=""
    crossorigin="anonymous"
    src="https://example.com/example.mp3"
  >
    Your browser does not support the HTML5 audio element.
  </audio>
</div>
`;
